web/src/upload.js

```diff
@@ -316,13 +316,41 @@ export function setup_upload(config) {
         },
     );
 
+    // To prevent overlay blinking, keep track of the number of
+    // drag events triggered. Ensures the overlay screen appears
+    // only when a file enters the drop zone and disappears either
+    // when the file is dropped or leaves the drop zone.
+    let drag_counter = 0;
+
     const $drag_drop_container = get_item("drag_drop_container", config);
-    $drag_drop_container.on("dragover", (event) => event.preventDefault());
-    $drag_drop_container.on("dragenter", (event) => event.preventDefault());
+    $drag_drop_container.on("dragover", (event) => {
+        event.preventDefault();
+        // Show the upload overlay during drag events.
+        $(".upload-overlay-container").addClass("upload-overlay-visibility");
+    });
+
+    $drag_drop_container.on("dragenter", (event) => {
+        event.preventDefault();
+        drag_counter += 1;
+        // Show the upload overlay during drag events.
+        $(".upload-overlay-container").addClass("upload-overlay-visibility");
+    });
+
+    $drag_drop_container.on("dragleave", (event) => {
+        event.preventDefault();
+        drag_counter -= 1;
```

Comment:
I'm wondering if these counters pass the "fidget test," where someone speedily brings a file in and out of the drag area rapidly? I haven't tried this, but I wonder if a technique like throttling or perhaps would work to head off the flicker that this aims to solve?

Reply:
It seems to work here but I will try the techniques as well that you mentioned.
[Screencast from 02-02-24 08:56:23 AM IST.webm](https://github.com/zulip/zulip/assets/90370535/3b257042-8b25-4c7d-9898-285f7844d289)